torch/testing/_internal/optests/autograd_registration.py

```diff
@@ -0,0 +1,135 @@
+    try:
+        prev = torch._C._get_autograd_fallback_mode()
+        torch._C._set_autograd_fallback_mode("nothing")
+        yield
+    finally:
+        torch._C._set_autograd_fallback_mode(prev)
+
+
+def autograd_registration_check(op, args, kwargs):
+    """Check if autograd was registered correctly (for the operator).
+
+    Operators should have "autograd support" registered directly to an
+    autograd dispatch key.
+    An incorrect registration may lead to unexpected silent incorrectness.
+    Note that this check won't catch all problems but will catch
+    the most common ones.
+
+    Example usage:
+        >>> x = torch.randn(3, requires_grad=True)
+        >>> autograd_registration_check(torch.ops.aten.sin.default, (x,), {})
+
+    It is important that you pass us at least one Tensor that has
+    requires_grad=True so that the autograd path gets exercised,
+    otherwise, this check may erroneously pass.
+
+    Here are some best practices if you do find your autograd is
+    registered incorrectly:
+    - If the operator is composite (i.e. consists of other PyTorch ops)
+      and you wish the operator to decompose and get autograd support
+      that way, then please register the implementation to
+      DispatchKey::CompositeImplicitAutograd
+    - If you're adding an autograd formula for the operator, the correct
+      thing to do is to register an autograd.Function to
+      DispatchKey::Autograd (preferred) or one of the
+      DispatchKey::Autograd<BACKEND> keys. It is NOT OK to register
+      an autograd.Function to a backend (e.g. CPU/CUDA) key.
+    - If your operator is non-differentiable, then you should register
+      an implementation to the Autograd key that uses
+      AutoDispatchBelowAutograd and re-invokes the operator.
+
+    """
+    assert isinstance(op, torch._ops.OpOverload)
+    # Implementation details
+    # -----------------------------------------------
+    # If an operator doesn't have an autograd kernel at an autograd key,
+    # and the operator does not return inputs as-is, then all of
+    # the outputs should have requires_grad=False before we apply
+    # special behaviors of our default autograd fallback.
+    # (The default autograd fallback may set requires_grad=True on output
+    # tensors in certain modes so that when they are backpropped through,
+    # they raise an error).
+    #
+    # Our strategy for detecting if an operator doesn't have an autograd
+    # kernel at the autograd key is:
+    # - set the autograd fallback mode to "nothing" (so it does not change
+    #   the required-gradness of outputs)
+    # - run the operator
+    # - Check if any outputs of the operator (that are not inputs) require
+    #   grad. This would only happen if the user calls regular PyTorch
+    #   operations in their backend key (this op should instead be
+    #   CompositeImplicitAutograd or not an op) or if the user invokes
+    #   an autograd.Function in the backend key.
```

Comment:
What if the op is partially a composite op and partially does non-pytorch operations? Are we sure we also want to error in that case? I guess in theory this means they could extract out a smaller custom operation. I'm not sure how often this happens in practice though.

If we wanted to test that the operator is "fully" composite that would probably mean having to actually validate the gradients are wrong, but maybe that is too complex to be worth?

Reply:
> What if the op is partially a composite op and partially does non-pytorch operations? Are we sure we also want to error in that case?

Yup. Ideally the user makes the custom op as small as possible.

> If we wanted to test that the operator is "fully" composite that would probably mean having to actually validate the gradients are wrong, but maybe that is too complex to be worth?

Yeah, this is meant to be a simple test.

Reply:
Ah makes sense. I don't know how common it is for folks to mix normal torch ops in with their custom ops, but if it is, it might be worth mentioning this case in the error/docs.